### PR TITLE
feat: implement ALP-RD compression

### LINKED ISSUE #10: Encoding: ALPRD
For floats that fail ALP encoding (e.g., logical "real" numbers rather than fixed-point decimals stored as FP)

### PR DESCRIPTION
**Pull request by a10y**
Fixes #10: Add ALP-RD compression.

Currently our only floating point compression algorithm is standard ALP, which targets floats/doubles that are originally decimal, and thus have some natural integer they can round to when you undo the exponent.

For science/math datasets, there are a lot of "real doubles", i.e. floating point numbers that use most/all of their available precision. These do not compress with standard ALP. The ALP paper authors had a solution for this called "ALP for 'Real' Doubles" / ALP-RD, which is implemented in this PR.

## Basics

The key insight of ALP-RD is that even for dense floating point numbers, within a column they often share the front bits (exponent + first few bits of mantissa). We try and find the best cut-point within the leftmost 16-bits.

There are generally a small number of unique values for the leftmost bits, so you can create a dictionary of fixed size (here we use the choice of 8 from the C++ implementation) which naturally bit-packs down to 3 bits. If you compress perfectly without exceptions, you can store 49 bits/value ~23% compression. In practice the amount varies. In the comments below you can see a test with the POI dataset referenced in the ALP paper, and we replicate their results of 55 and 56 bits/value respectively.

## List of changes

* Reorganized the `vortex-alp` crate. I created two top-level modules, `alp` and alp_rd`, and moved the previous implementation into the `alp` module
* Added new `ALPRDArray` in the `alp_rd` module. It supports both f32 and f64, and all major compute functions are implemented (save for `MaybeCompareFn` and the Accessors I will file an issue to implement these in a FLUP if alright, this PR is already quite large)
* Added corresponding `ALPRDCompressor` and wired the CompressorRef everywhere I could find ALPCompressor
* New benchmark for RD compression in the existing ALP benchmarks suite

### CODE REVIEW

**Comment by a10y** [comment]
Some Q's:

1. For ALP-RD, we store a small dictionary (<= 16 bytes) for the left-parts. Should that be stored as a child or as metadata on the array?
2. ALP-RD from the paper prescribes specific compression algorithms for each of its subcomponents (fused dict+FL for left-parts, FL bit-pack for right-parts). Should we implement those, or should we just let it cascade in the compressor?

Separately just an observation, but ALP from the paper recommends having one pair of exponents per vector, rather than for the entire array like we do now.

## EDIT: answers

1. Store in metadata
2. Can bit-pack the left/right side explicitly

File: encodings/alp/src/alp_rd/mod.rs [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+
+        for v in doubles.iter().copied() {
+            right_parts.push(T::to_bits(v) & right_mask);
+            left_parts.push(T::to_u16(T::to_bits(v).shr(self.right_bit_width as _)));
+        }
+
+        // dict-encode the left-parts, keeping track of exceptions
+        for (idx, left) in left_parts.iter_mut().enumerate() {
+            // TODO: revisit if we need to change the branch order for perf.
+            if let Some(code) = self.codes.iter().position(|v| *v == *left) {
```
Comment (a10y) [inline encodings/alp/src/alp_rd/mod.rs:192]: I had originally used HashMap for this, like the C++ code does, but it turns out that doing linear search on a small fixed-size array is considerably faster (~5x) than doing hashmap lookups

**Comment by a10y** [comment]
I implemented a small test using the POI [Kaggle dataset](https://www.kaggle.com/datasets/ehallmar/points-of-interest-poi-database) referenced from the paper, and I was able to replicate the compression ration results.


image


```
/Users/aduffy/.cargo/bin/cargo run --color=always --bin compress_poi --manifest-path /Volumes/Code/vortex/bench-vortex/Cargo.toml
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.16s
     Running `target/debug/compress_poi`
reading with schema: Struct(StructDType { names: ["name", "latitude_radian", "longitude_radian", "num_links", "links", "num_categories", "categories"], dtypes: [Utf8(Nullable), Primitive(F64, Nullable), Primitive(F64, Nullable), Primitive(I64, Nullable), Utf8(Nullable), Primitive(I64, Nullable), Utf8(Nullable)] }, NonNullable)
raw POI dataset
root: vortex.struct(0x04)({latitude_radian=f64?, longitude_radian=f64?}, len=424205) nbytes=6.79 MB (100.00%)
  metadata: StructMetadata { length: 424205, validity: NonNullable }
  "latitude_radian": vortex.primitive(0x03)(f64?, len=424205) nbytes=3.39 MB (50.00%)
    metadata: PrimitiveMetadata { validity: AllValid }
    buffer: 3.39 MB
  "longitude_radian": vortex.primitive(0x03)(f64?, len=424205) nbytes=3.39 MB (50.00%)
    metadata: PrimitiveMetadata { validity: AllValid }
    buffer: 3.39 MB

Compressed POI data
root: vortex.struct(0x04)({latitude_radian=f64?, longitude_radian=f64?}, len=424205) nbytes=5.95 MB (100.00%)
  metadata: StructMetadata { length: 424205, validity: NonNullable }
  "latitude_radian": vortex.alprd(0x1e)(f64?, len=424205) nbytes=2.95 MB (49.66%)
    metadata: ALPRDMetadata { is_f32: false, right_bit_width: 52, dict_len: 8, dict: [1022, 1021, 3069, 1020, 1023, 3070, 3068, 3071], left_parts_dtype: Primitive(U16, Nullable), has_exceptions: true }
    left_parts: fastlanes.bitpacked(0x15)(u16?, len=424205) nbytes=159.08 kB (2.68%)
      metadata: BitPackedMetadata { validity: AllValid, bit_width: 3, offset: 0, length: 424205, has_patches: false }
      buffer: 159.36 kB
    right_parts: fastlanes.bitpacked(0x15)(u64?, len=424205) nbytes=2.76 MB (46.38%)
      metadata: BitPackedMetadata { validity: AllValid, bit_width: 52, offset: 0, length: 424205, has_patches: false }
      buffer: 2.76 MB
    left_parts_exceptions: vortex.sparse(0x08)(u16?, len=424205) nbytes=36.42 kB (0.61%)
      metadata: SparseMetadata { indices_dtype: Primitive(U64, NonNullable), indices_offset: 0, indices_len: 8325, len: 424205, fill_value: Scalar { dtype: Primitive(U16, Nullable), value: Null } }
      indices: fastlanes.bitpacked(0x15)(u64, len=8325) nbytes=19.77 kB (0.33%)
        metadata: BitPackedMetadata { validity: NonNullable, bit_width: 19, offset: 0, length: 8325, has_patches: false }
        buffer: 21.89 kB
      values: vortex.primitive(0x03)(u16?, len=8325) nbytes=16.65 kB (0.28%)
        metadata: PrimitiveMetadata { validity: AllValid }
        buffer: 16.65 kB
  "longitude_radian": vortex.alprd(0x1e)(f64?, len=424205) nbytes=2.99 MB (50.34%)
    metadata: ALPRDMetadata { is_f32: false, right_bit_width: 53, dict_len: 8, dict: [1535, 510, 511, 512, 1536, 509, 1533, 1534], left_parts_dtype: Primitive(U16, Nullable), has_exceptions: true }
    left_parts: fastlanes.bitpacked(0x15)(u16?, len=424205) nbytes=159.08 kB (2.68%)
      metadata: BitPackedMetadata { validity: AllValid, bit_width: 3, offset: 0, length: 424205, has_patches: false }
      buffer: 159.36 kB
    right_parts: fastlanes.bitpacked(0x15)(u64?, len=424205) nbytes=2.81 MB (47.27%)
      metadata: BitPackedMetadata { validity: AllValid, bit_width: 53, offset: 0, length: 424205, has_patches: false }
      buffer: 2.82 MB
    left_parts_exceptions: vortex.sparse(0x08)(u16?, len=424205) nbytes=23.37 kB (0.39%)
      metadata: SparseMetadata { indices_dtype: Primitive(U64, NonNullable), indices_offset: 0, indices_len: 5342, len: 424205, fill_value: Scalar { dtype: Primitive(U16, Nullable), value: Null } }
      indices: fastlanes.bitpacked(0x15)(u64, len=5342) nbytes=12.69 kB (0.21%)
        metadata: BitPackedMetadata { validity: NonNullable, bit_width: 19, offset: 0, length: 5342, has_patches: false }
        buffer: 14.59 kB
      values: vortex.primitive(0x03)(u16?, len=5342) nbytes=10.68 kB (0.18%)
        metadata: PrimitiveMetadata { validity: AllValid }
        buffer: 10.68 kB
```

We can see that our bits-per-value are roughly 55 for `latitude_radians` and 56 for `longitude_radians`:

* latitude = 55.6 bits per pixel
* longitude: 56.4

This nets us an overall compression ratio of ~12.5

Screenshot 2024-10-01 at 3 41 27 PM
%

File: bench-vortex/src/reader.rs [hunk 2]
```diff
@@ -89,7 +89,7 @@ pub async fn rewrite_parquet_as_vortex<W: VortexWrite>(
     Ok(())
 }
 
-pub fn read_parquet_to_vortex(parquet_path: &Path) -> VortexResult<ChunkedArray> {
+pub fn read_parquet_to_vortex<P: AsRef<Path>>(parquet_path: P) -> VortexResult<ChunkedArray> {
```
Comment (a10y) [inline bench-vortex/src/reader.rs:92]: ergonomics

File: encodings/alp/src/alp/mod.rs [hunk 3]
```diff
@@ -19,7 +26,14 @@ impl Display for Exponents {
     }
 }
 
-pub trait ALPFloat: Float + Display + 'static {
+mod private {
```
Comment (a10y) [inline encodings/alp/src/alp/mod.rs:29]: in theory this was previously extensible, but we probably want to constrain it

File: encodings/alp/src/alp_rd/mod.rs [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+        };
+
+        // Bit-pack the dict-encoded left-parts
+        // Bit-pack the right-parts
+        // SparseArray for exceptions.
+        let exceptions = (!exceptions_pos.is_empty()).then(|| {
+            let max_exc_pos = exceptions_pos.last().copied().unwrap_or_default();
+            // Add one to get next power of two as well here.
+            // If we're going to be doing more of this, it just works.
+            let bw = (max_exc_pos + 1).next_power_of_two().ilog2() as usize;
```
Comment (a10y) [inline encodings/alp/src/alp_rd/mod.rs]: undo this

File: encodings/alp/src/alp_rd/mod.rs [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+
+            let exc_pos_array = PrimitiveArray::from(exceptions_pos);
+            let packed_pos = BitPackedArray::encode(exc_pos_array.as_ref(), bw)
+                .vortex_unwrap()
+                .into_array();
```
Comment (a10y) [inline encodings/alp/src/alp_rd/mod.rs]: use bitpack_encode_unchecked here as well

File: encodings/alp/src/alp_rd/mod.rs [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+            &self.codes,
+            packed_right,
+            self.right_bit_width,
+            exceptions,
+        )
+        .vortex_expect("ALPRDArray construction in encode")
+    }
+}
+
+// Only applies for F64.
```
Comment (a10y) [inline encodings/alp/src/alp_rd/mod.rs]: old comment. replace with real doc comment

File: encodings/alp/src/alp/mod.rs (hunk 3 above)
Comment (robert3005) [reply encodings/alp/src/alp/mod.rs:29]: poor f16 not considered in the paper. Anyway this is the right thing to do

**Comment by robert3005** [comment]
one note - don't bother with accessors we decided that we likely need to change them

File: encodings/alp/src/alp/mod.rs (hunk 3 above)
Comment (a10y) [reply encodings/alp/src/alp/mod.rs:29]: ah i too forgot about f16. I suppose that we probably want special compressors for things like bf16

Comment (robert3005) [reply encodings/alp/src/alp/mod.rs:29]: Yes, this is just a comment for future readers. The paper only talked about the common float types

File: bench-vortex/src/lib.rs [hunk 7]
```diff
@@ -54,6 +55,7 @@ lazy_static! {
 lazy_static! {
     pub static ref COMPRESSORS: HashSet<CompressorRef<'static>> = [
         &ALPCompressor as CompressorRef<'static>,
+        &ALPRDCompressor as CompressorRef<'static>,
```
Comment (robert3005) [inline bench-vortex/src/lib.rs]: no need for `as` here, first one is just the hint for compiler

File: vortex-dtype/src/dtype.rs [hunk 8]
```diff
@@ -112,6 +112,13 @@ impl DType {
             _ => None,
         }
     }
+
+    pub fn ptype(&self) -> Option<PType> {
```
Comment (robert3005) [inline vortex-dtype/src/dtype.rs]: There's `TryFrom<&DType> for PType`, should decide on one of them

File: vortex-sampling-compressor/src/compressors/alp_rd.rs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+        let encoded = alp_rd_encoder.encode(&primitive).into_array();
+        Ok(CompressedArray::new(
+            encoded,
+            Some(CompressionTree::new_with_metadata(self, vec![], encoder)),
+        ))
+    }
+
+    fn used_encodings(&self) -> HashSet<EncodingRef> {
+        HashSet::from([&ALPRDEncoding as EncodingRef])
```
Comment (robert3005) [inline vortex-sampling-compressor/src/compressors/alp_rd.rs]: you should return `BitPackedEncoding` here as well

File: encodings/alp/src/alp_rd/compute/slice.rs [hunk 10]
```diff
@@ -0,0 +1,59 @@
+use vortex::compute::{slice, SliceFn};
+use vortex::{Array, ArrayDType, IntoArray};
+use vortex_error::VortexResult;
+
+use crate::ALPRDArray;
+
+impl SliceFn for ALPRDArray {
+    fn slice(&self, start: usize, stop: usize) -> VortexResult<Array> {
+        let left_parts_exceptions = match self.left_parts_exceptions() {
```
Comment (robert3005) [inline encodings/alp/src/alp_rd/compute/slice.rs]: make it a map call?

File: encodings/alp/src/alp_rd/compute/filter.rs [hunk 11]
```diff
@@ -0,0 +1,60 @@
+use vortex::compute::{filter, FilterFn};
+use vortex::{Array, ArrayDType, IntoArray};
+use vortex_error::VortexResult;
+
+use crate::ALPRDArray;
+
+impl FilterFn for ALPRDArray {
+    fn filter(&self, predicate: &Array) -> VortexResult<Array> {
+        let left_parts_exceptions = match self.left_parts_exceptions() {
```
Comment (robert3005) [inline encodings/alp/src/alp_rd/compute/filter.rs]: make it a map call

File: encodings/alp/src/alp_rd/array.rs [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+use vortex_dtype::{DType, PType};
+use vortex_error::{vortex_bail, VortexExpect, VortexResult};
+
+use crate::alp_rd::alp_rd_decode;
+
+impl_encoding!("vortex.alprd", ids::ALP_RD, ALPRD);
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ALPRDMetadata {
+    is_f32: bool,
```
Comment (robert3005) [inline encodings/alp/src/alp_rd/array.rs]: you have array dtype for this

File: encodings/alp/Cargo.toml [hunk 13]
```diff
@@ -17,6 +17,8 @@ readme = { workspace = true }
 workspace = true
 
 [dependencies]
+fastlanes = { workspace = true }
```
Comment (a10y) [inline encodings/alp/Cargo.toml]: Unused

File: encodings/alp/src/alp_rd/array.rs (hunk 12 above)
Comment (robert3005) [reply encodings/alp/src/alp_rd/array.rs]: Could just add this as a small function and remove this bit of metadata

**Review by robert3005 — APPROVED** [review]
some small changes

File: bench-vortex/src/lib.rs (hunk 7 above)
Comment (a10y) [reply bench-vortex/src/lib.rs]: yea that was copy-paste error on my part

**Comment by lwwmanning** [comment]
🥳

**Pull request merged**